lib/WhereClause.php

```diff
@@ -0,0 +1,342 @@
+     * @param array<mixed>         $values Array of values for the field names. This is used
+     *                                     to determine what kind of bind marker to use: =?, IN(?), IS NULL
+     * @param array<string,string> $map    A hash of "mapped_column_name" => "real_column_name"
+     */
+    public static function from_underscored_string(
+        Connection $connection, string $name, array $values = [], array $map = []): WhereClause
+    {
+        $expression = '';
+
+        $num_values = count((array) $values);
+        $conditionValues = [];
+
+        $parts = SQLBuilder::underscored_string_to_parts($name);
+
+        for ($i = 0, $j = 0, $n = count($parts); $i < $n; $i += 2, ++$j) {
+            if ($i >= 2) {
+                $res = preg_replace(['/_and_/i', '/_or_/i'], [' AND ', ' OR '], $parts[$i - 1]);
+                assert(is_string($res));
+                $expression .= $res;
+            }
+
+            if ($j < $num_values) {
+                if (!is_null($values[$j])) {
+                    $bind = is_array($values[$j]) ? ' IN(?)' : '=?';
+                    $conditionValues[] = $values[$j];
+                } else {
+                    $bind = ' IS NULL';
+                }
+            } else {
+                $bind = ' IS NULL';
+            }
+
+            // map to correct name if $map was supplied
+            $name = $map && isset($map[$parts[$i]]) ? $map[$parts[$i]] : $parts[$i];
+
+            $expression .= $connection->quote_name($name) . $bind;
+        }
+
+        return new WhereClause($expression, $conditionValues);
+    }
+
+    public static function from_arg(mixed $arg): WhereClause
+    {
+        // user passed in a string, a hash, or an array consisting of a string and values
+        if (is_string($arg) || is_hash($arg)) {
+            $expression = new WhereClause($arg, []);
+        } else {
+            $expression = new WhereClause($arg[0], array_slice($arg, 1));
+        }
+
+        return $expression;
+    }
+
+    /**
+     * Replaces any aliases used in a hash based condition.
+     *
+     * @param array<string, string> $hash A hash
+     * @param array<string, string> $map  Hash of used_name => real_name
+     *
+     * @return array<string, string> Array with any aliases replaced with their read field name
+     */
+    private function map_names(array &$hash, array &$map): array
+    {
+        $ret = [];
+
+        foreach ($hash as $name => &$value) {
+            if (array_key_exists($name, $map)) {
+                $name = $map[$name];
+            }
+
+            $ret[$name] = $value;
+        }
+
+        return $ret;
+    }
+
+    /**
+     * @param Attributes $hash
+     *
+     * @return array<mixed>
+     */
+    private function build_sql_from_hash(array $hash, string $prependTableName, string $glue = ' AND '): array
+    {
+        $sql = $g = '';
+
+        $table = !empty($prependTableName) ? $this->connection->quote_name($prependTableName) : '';
```

Comment:
Crash if $this->connection == null.

I think $this->connection should always be set in the class constructor and then get rid of getter/setter for connection. $this->connection should be immutable across an HTTP call, so you can use the [Prototype pattern](https://refactoring.guru/design-patterns/prototype) to set it once and then clone WhereClauses.

I would prefer though the use of a Dependency Injector like [Dice](https://github.com/Level-2/Dice) which I for my projects and recommend it. The connection would then be a singleton across the whole app.

Reply:
> $this->connection should always be set in the class constructor

Well, that's sort of how it was when `Expression` was only being instantiated inside `SQLBuilder`, but now that I'm spawning them willy-nilly all over the place I don't want to deal with connections until we're absolutely ready to start talking to the database. 

But you're right, I was well aware of the situation here and was just being lazy. I've removed the `$connection` member and all the setter/getter business and now just pass a `Connection` around as needed. 6737374